File: Go/games/flappy.go
package games

import (
	"fmt"
	"image/color"
	"math/rand"

	"github.com/hajimehoshi/ebiten/v2"
	"github.com/hajimehoshi/ebiten/v2/inpututil"
	"github.com/hajimehoshi/ebiten/v2/text"
	"github.com/hajimehoshi/ebiten/v2/vector"
)

type Pipe struct {
	X      int32
	GapY   int32
	Passed bool
}

type FlappyGame struct {
	width      int32
	height     int32
	offsetX    int32
	offsetY    int32
	birdY      float32
	birdVY     float32
	birdRadius int32
	birdX      int32
	gravity    float32
	jumpStr    float32
	pipes      []Pipe
	pipeWidth  int32
	pipeGap    int32
	pipeSpeed  int32
	score      int32
	gameOver   bool
	pipeTimer  int32
	pipeDelay  int32
}

func NewFlappyGame() *FlappyGame {
	g := &FlappyGame{
		width:      400,
		height:     600,
		birdRadius: 15,
		birdX:      80,
		gravity:    0.6,
		jumpStr:    -11,
		pipeWidth:  60,
		pipeGap:    160,
		pipeSpeed:  3,
		pipeDelay:  90,
	}
	g.offsetX = (ScreenWidth - g.width) / 2
	g.offsetY = (ScreenHeight - g.height) / 2
	g.Reset()
	return g
}

func (g *FlappyGame) Reset() {
	g.birdY = float32(g.height / 2)
	g.birdVY = 0
	g.pipes = nil
	g.score = 0
	g.gameOver = false
	g.pipeTimer = 0
}

func (g *FlappyGame) addPipe() {
	gapY := int32(rand.Intn(int(g.height-80-g.pipeGap))) + 40
	g.pipes = append(g.pipes, Pipe{X: g.width, GapY: gapY, Passed: false})
}

func (g *FlappyGame) Update() error {
	if inpututil.IsKeyJustPressed(ebiten.KeyEscape) {
		return ebiten.Termination
	}
	if inpututil.IsKeyJustPressed(ebiten.KeyR) {
		g.Reset()
		return nil
	}
	if inpututil.IsKeyJustPressed(ebiten.KeySpace) ||
		inpututil.IsMouseButtonJustPressed(ebiten.MouseButtonLeft) {
		if !g.gameOver {
			g.birdVY = g.jumpStr
		}
	}

	if g.gameOver {
		return nil
	}

	g.birdVY += g.gravity
	g.birdY += g.birdVY

	if g.birdY-float32(g.birdRadius) < 0 {
		g.birdY = float32(g.birdRadius)
		g.birdVY = 0
	} else if g.birdY+float32(g.birdRadius) > float32(g.height) {
		g.gameOver = true
		return nil
	}

	g.pipeTimer++
	if g.pipeTimer >= g.pipeDelay {
		g.pipeTimer = 0
		g.addPipe()
	}

	for i := range g.pipes {
		g.pipes[i].X -= g.pipeSpeed
		if !g.pipes[i].Passed && g.pipes[i].X+g.pipeWidth < g.birdX {
			g.pipes[i].Passed = true
			g.score++
		}
		if g.pipes[i].X < g.birdX+int32(g.birdRadius) &&
			g.pipes[i].X+g.pipeWidth > g.birdX-int32(g.birdRadius) {
			if g.birdY-float32(g.birdRadius) < float32(g.pipes[i].GapY) ||
				g.birdY+float32(g.birdRadius) > float32(g.pipes[i].GapY+g.pipeGap) {
				g.gameOver = true
				return nil
			}
		}
	}

	newPipes := make([]Pipe, 0)
	for _, p := range g.pipes {
		if p.X+g.pipeWidth > 0 {
			newPipes = append(newPipes, p)
		}
	}
	g.pipes = newPipes
	return nil
}

func (g *FlappyGame) Draw(screen *ebiten.Image) {
	screen.Fill(ColorBg)

	DrawRoundedRect(screen, float32(g.offsetX), float32(g.offsetY),
		float32(g.width), float32(g.height), 12, color.RGBA{135, 206, 235, 255})

	bx := float32(g.offsetX + g.birdX)
	by := float32(g.offsetY) + g.birdY
	vector.DrawFilledCircle(screen, bx, by, float32(g.birdRadius), color.RGBA{255, 255, 50, 255}, true)
	vector.DrawFilledRect(screen, bx+8, by-5, 4, 4, color.Black, true)
	vector.DrawFilledRect(screen, bx+10, by-7, 2, 2, color.White, true)

	for _, pipe := range g.pipes {
		topRect := float32(g.offsetY)
		topH := float32(pipe.GapY)
		DrawRoundedRect(screen, float32(g.offsetX+pipe.X), topRect,
			float32(g.pipeWidth), topH, 0, color.RGBA{40, 200, 40, 255})
		DrawRoundedRect(screen, float32(g.offsetX+pipe.X-6),
			topRect+topH-22, float32(g.pipeWidth+12), 22, 6,
			color.RGBA{30, 160, 30, 255})

		bottomY := float32(g.offsetY + pipe.GapY + g.pipeGap)
		bottomH := float32(g.height) - float32(pipe.GapY+g.pipeGap)
		DrawRoundedRect(screen, float32(g.offsetX+pipe.X), bottomY,
			float32(g.pipeWidth), bottomH, 0, color.RGBA{40, 200, 40, 255})
		DrawRoundedRect(screen, float32(g.offsetX+pipe.X-6), bottomY,
			float32(g.pipeWidth+12), 22, 6, color.RGBA{30, 160, 30, 255})
	}

	if FontBig != nil {
		s := fmt.Sprint(g.score)
		b := text.BoundString(FontBig, s)
		text.Draw(screen, s, FontBig,
			int(g.offsetX)+int(g.width)/2-b.Dx()/2,
			int(g.offsetY)+50, ColorWhite)
	}

	if g.gameOver {
		DrawOverlay(screen, int(g.offsetX), int(g.offsetY), int(g.width), int(g.height))
		msg := "GAME OVER"
		b := text.BoundString(FontBig, msg)
		text.Draw(screen, msg, FontBig,
			int(g.offsetX)+int(g.width)/2-b.Dx()/2,
			int(g.offsetY)+int(g.height)/2-30, ColorWhite)
		msg2 := "Press R to restart"
		b2 := text.BoundString(FontSmall, msg2)
		text.Draw(screen, msg2, FontSmall,
			int(g.offsetX)+int(g.width)/2-b2.Dx()/2,
			int(g.offsetY)+int(g.height)/2+20, ColorWhite)
	}
}

func (g *FlappyGame) Layout(outsideWidth, outsideHeight int) (int, int) {
	return ScreenWidth, ScreenHeight
}

File: Rust/src/games/flappy.rs
use crate::common::*;
use macroquad::prelude::*;
use ::rand::Rng;

const W: f32 = 400.0;
const H: f32 = 600.0;
const BIRD_X: f32 = 80.0;

struct Pipe {
    x: f32,
    gap_y: f32,
    passed: bool,
}

pub struct FlappyGame {
    bird_y: f32,
    bird_vy: f32,
    bird_radius: f32,
    gravity: f32,
    jump_strength: f32,
    pipes: Vec<Pipe>,
    pipe_width: f32,
    pipe_gap: f32,
    pipe_speed: f32,
    score: i32,
    game_over: bool,
    pipe_timer: i32,
    pipe_delay: i32,
    offset: (f32, f32),
}

impl FlappyGame {
    pub fn new() -> Self {
        let mut g = Self {
            bird_y: H / 2.0,
            bird_vy: 0.0,
            bird_radius: 15.0,
            gravity: 0.2,
            jump_strength: -7.0,
            pipes: Vec::new(),
            pipe_width: 60.0,
            pipe_gap: 160.0,
            pipe_speed: 2.0,
            score: 0,
            game_over: false,
            pipe_timer: 0,
            pipe_delay: 90,
            offset: centered_offset(W, H),
        };
        g.reset();
        g
    }

    fn reset(&mut self) {
        self.bird_y = H / 2.0;
        self.bird_vy = 0.0;
        self.pipes.clear();
        self.score = 0;
        self.game_over = false;
        self.pipe_timer = 0;
    }

    fn add_pipe(&mut self) {
        let mut rng = ::rand::thread_rng();
        let gap_y = rng.gen_range(80.0..=(H - 80.0 - self.pipe_gap));
        self.pipes.push(Pipe {
            x: W,
            gap_y,
            passed: false,
        });
    }
}

impl Game for FlappyGame {
    fn update(&mut self) -> bool {
        // ---- 事件 ----
        if is_key_pressed(KeyCode::Escape) {
            return false;
        }
        if is_key_pressed(KeyCode::R) {
            self.reset();
        }
        if is_key_pressed(KeyCode::Space) && !self.game_over {
            self.bird_vy = self.jump_strength;
        }
        if is_mouse_button_pressed(MouseButton::Left) && !self.game_over {
            self.bird_vy = self.jump_strength;
        }

        if self.game_over {
            return true;
        }

        // ---- 逻辑 ----
        self.bird_vy += self.gravity;
        self.bird_y += self.bird_vy;

        if self.bird_y - self.bird_radius < 0.0 {
            self.bird_y = self.bird_radius;
            self.bird_vy = 0.0;
        } else if self.bird_y + self.bird_radius > H {
            self.game_over = true;
            return true;
        }

        self.pipe_timer += 1;
        if self.pipe_timer >= self.pipe_delay {
            self.pipe_timer = 0;
            self.add_pipe();
        }

        let mut remove_idx: Vec<usize> = Vec::new();
        for i in 0..self.pipes.len() {
            self.pipes[i].x -= self.pipe_speed;

            if !self.pipes[i].passed && self.pipes[i].x + self.pipe_width < BIRD_X {
                self.pipes[i].passed = true;
                self.score += 1;
            }

            // 碰撞
            let p = &self.pipes[i];
            if p.x < BIRD_X + self.bird_radius && p.x + self.pipe_width > BIRD_X - self.bird_radius {
                if self.bird_y - self.bird_radius < p.gap_y
                    || self.bird_y + self.bird_radius > p.gap_y + self.pipe_gap
                {
                    self.game_over = true;
                    return true;
                }
            }

            if p.x + self.pipe_width < 0.0 {
                remove_idx.push(i);
            }
        }
        for i in remove_idx.into_iter().rev() {
            self.pipes.remove(i);
        }

        true
    }

    fn draw(&mut self) {
        clear_background(COLOR_BG);
        let (ox, oy) = self.offset;

        // 天空
        draw_rectangle(ox, oy, W, H, Color::new(0.529, 0.808, 0.922, 1.0));

        // 鸟
        let bx = ox + BIRD_X;
        let by = oy + self.bird_y;
        draw_circle(bx, by, self.bird_radius, Color::new(1.0, 1.0, 0.196, 1.0));
        draw_circle(bx + 8.0, by - 5.0, 4.0, BLACK);
        draw_circle(bx + 10.0, by - 7.0, 2.0, WHITE);

        // 管道
        for p in &self.pipes {
            let top = Rect::new(ox + p.x, oy, self.pipe_width, p.gap_y);
            draw_rounded_rect(top.x, top.y, top.w, top.h, 6.0, Color::new(0.157, 0.784, 0.157, 1.0));
            draw_rounded_rect(
                ox + p.x - 6.0,
                oy + p.gap_y - 22.0,
                self.pipe_width + 12.0,
                22.0,
                6.0,
                Color::new(0.118, 0.627, 0.118, 1.0),
            );

            let bottom_y = p.gap_y + self.pipe_gap;
            draw_rounded_rect(
                ox + p.x,
                oy + bottom_y,
                self.pipe_width,
                H - bottom_y,
                6.0,
                Color::new(0.157, 0.784, 0.157, 1.0),
            );
            draw_rounded_rect(
                ox + p.x - 6.0,
                oy + bottom_y,
                self.pipe_width + 12.0,
                22.0,
                6.0,
                Color::new(0.118, 0.627, 0.118, 1.0),
            );
        }

        // 分数
        draw_text_centered(
            &self.score.to_string(),
            ox + W / 2.0,
            oy + 50.0,
            40,
            WHITE,
        );

        // Game Over
        if self.game_over {
            draw_overlay(ox, oy, W, H, 0.6, BLACK);
            draw_text_centered("GAME OVER", ox + W / 2.0, oy + H / 2.0 - 30.0, 40, WHITE);
            draw_text_centered("Press R to restart", ox + W / 2.0, oy + H / 2.0 + 20.0, 26, WHITE);
        }
    }
}
